File: src/BusinessLogic/Handlers/InlineQueryHandlers/DubsInlineQueryHandler.cs
using Bounan.Bot.BusinessLogic.CommandDto;
using Bounan.Bot.BusinessLogic.Constants;
using Bounan.Bot.BusinessLogic.Helpers;
using Bounan.Bot.TelegramBot.Telegram;
using Bounan.LoanApi.Interfaces;
using Microsoft.Extensions.Logging;
using Telegram.Bot.Requests;
using Telegram.Bot.Types;
using Telegram.Bot.Types.InlineQueryResults;

namespace Bounan.Bot.BusinessLogic.Handlers.InlineQueryHandlers;

public class DubsInlineQueryHandler(
    ILogger<DubsInlineQueryHandler> logger,
    ILoanApiComClient botLoanApiClient) : IInlineQueryHandler
{
    public static bool CanHandle(InlineQuery item) => item.Query.StartsWith(DubsCommandDto.Command);

    public async Task<AnswerInlineQueryRequest> HandleAsync(
        InlineQuery inlineQuery,
        CancellationToken cancellationToken)
    {
        logger.LogInformation("Handling dubs inline query {Query}", inlineQuery.Query);

        var commandDto = CommandConvert.DeserializeCommand<DubsCommandDto>(inlineQuery.Query);
        if (commandDto is null)
        {
            logger.LogWarning("Failed to deserialize command {Command}", inlineQuery.Query);
            return TelegramInlineItemsHelpers.EmptyResults(inlineQuery.Id);
        }

        var searchResults = await botLoanApiClient.SearchAsync(commandDto.MyAnimeListId, cancellationToken);
        if (searchResults is null or { Count: 0 })
        {
            logger.LogInformation("No dubs for {MyAnimeListId}", commandDto.MyAnimeListId);
            return TelegramInlineItemsHelpers.ConstResult(inlineQuery.Id, KnownInlineAnswers.AnimeUnavailable);
        }

        var uniqueDubs = searchResults
            .OrderBy(item => item.Dub)
            .ThenBy(item => item.Episode)
            .DistinctBy(item => item.Dub);
        logger.LogInformation("Got dubs for {MyAnimeListId}: {Dubs}", commandDto.MyAnimeListId, uniqueDubs);

        var results = uniqueDubs
            .Select(item =>
            {
                var watchCommandDto = new WatchCommandDto
                {
                    MyAnimeListId = commandDto.MyAnimeListId,
                    Dub = AnimeHelpers.DubToKey(item.Dub),
                    Episode = item.Episode,
                };

                var serializedCommand = CommandConvert.SerializeCommand(watchCommandDto);

                return new InlineQueryResultArticle(
                    item.Dub,
                    item.Dub,
                    new InputTextMessageContent(serializedCommand));
            })
            .ToArray();

        logger.LogInformation("Returning {Count} dubs for {MyAnimeListId}", results.Length, commandDto.MyAnimeListId);
        return new AnswerInlineQueryRequest(inlineQuery.Id, results);
    }
}

File: src/handlers/on-webhook/event-handlers/inline-query-handlers/dubs-inline-query-handler.ts
import { InlineQuery, InlineQueryResultArticle } from '@lightweight-clients/telegram-bot-api-lightweight-client';

import { getDubs } from '../../../../api-clients/cached-loan-api-client';
import { dubToKey } from '../../../../shared/helpers/dub-to-key';
import { getStudioLogoUrl } from '../../../../shared/studio-logos-provider/studio-logos-provider';
import { DubsCommandDto, WatchCommandDto } from '../../command-dtos';
import { KnownInlineAnswers } from '../../constants/known-inline-answers';
import { InlineQueryHandler } from '../query-handler';

const canHandle = (inlineQuery: InlineQuery): boolean => inlineQuery.query?.startsWith(DubsCommandDto.Command) ?? false;

const handler: InlineQueryHandler = async (inlineQuery) => {
    console.log('Handling dubs inline query {Query}', inlineQuery.query);

    const commandDto = DubsCommandDto.fromPayload(inlineQuery.query) as DubsCommandDto;
    if (!commandDto) {
        console.warn('Failed to deserialize command', inlineQuery.query);
        return [];
    }

    const uniqueDubs = await getDubs(commandDto.myAnimeListId);
    console.log('Got dubs for {MyAnimeListId}: {Dubs}', commandDto.myAnimeListId, uniqueDubs);
    if (uniqueDubs.length === 0) {
        return [{
            type: 'article',
            id: KnownInlineAnswers.AnimeUnavailable,
            title: KnownInlineAnswers.AnimeUnavailable,
            input_message_content: {
                message_text: KnownInlineAnswers.AnimeUnavailable,
            },
        }];
    }

    const results: InlineQueryResultArticle[] = uniqueDubs
        .sort((a, b) => a.name.localeCompare(b.name))
        .map(item => ({
            type: 'article',
            id: item.name,
            title: item.name,
            thumbnail_url: getStudioLogoUrl(item.name),
            input_message_content: {
                message_text: new WatchCommandDto(
                    commandDto.myAnimeListId,
                    dubToKey(item.name),
                    item.firstEpisode,
                ).toString(),
            },
        }));

    console.log('Returning {Count} dubs for {MyAnimeListId}', results.length, commandDto.myAnimeListId);
    return results;
}

export const dubsInlineQueryHandler = {
    canHandle,
    handler,
};
